src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -5594,3 +5600,47 @@ ExtractParametersFromParamList(ParamListInfo paramListInfo,
 																   parameterData->value);
 	}
 }
+
+
+/*
+ * Delegated functions which are not part of another transaction are naturally safe,
+ * but if it's part of an outer transaction (a 2PC), we should ensure that the current
+ * task will not do remote(connection) activity to ensure the transactional integrity
+ * of the current participating 2PC.
+ */
+static void
+EnsureDelegatedFunctionSafeExecution(List *taskList)
+{
+	if (taskList == NIL)
+	{
+		/* we allow executions with an empty task list (e.g. WHERE false) */
+		return;
+	}
+
+	if (!IsCitusInitiatedRemoteBackend())
+	{
+		/* application-initiated backends coordinate the 2PC, need no restrictions */
+		return;
+	}
+
+	if (!MyBackendIsInDisributedTransaction())
+	{
+		/* if this backend is not participating in a 2PC, nothing to check */
+		return;
+	}
+
+	DistributedTransactionId *transactionId = GetCurrentDistributedTransactionId();
+	if (transactionId->initiatorNodeIdentifier == GetLocalGroupId())
+	{
+		/* Is the distributed transaction (ID) initiated on this node? */
```

Comment:
Why do we have this check? Even if a node connect to itself, we should not allow connections to other nodes from that session.

Reply:
A Citus-backend can't have a distributed transaction started on its own, it can be a participant. I removed the redundant check.

Reply:
ping

Reply:
I removed, but it broke ton of legacy tests, to confirm the case I put it back, still investigating.

Reply:
Here is a simple repro where we are hitting 

**psql:/tmp/proc:29: ERROR:  cannot access other workers from operations on shards**

```CREATE TABLE test_proc_table (a int);

create or replace procedure proc_pushdown(dist_key integer, OUT created int4[], OUT res_out text)
language plpgsql
as $$
DECLARE                                                                                                                                                          res INT := 0;
begin
    INSERT INTO test_proc_table VALUES (dist_key);
    SELECT count(*) INTO res FROM test_proc_table;
    created := created || res;
    PERFORM array_prepend(res, created);
    res_out := res::text;
    commit;
end;$$;                                                                                                                                                      

SELECT create_distributed_table('test_proc_table', 'a');
SELECT create_distributed_function('proc_pushdown(integer)', 'dist_key', 'test_proc_table' );                                                                
-- make sure that metadata is synced
SELECT bool_and(hasmetadata) FROM pg_dist_node WHERE nodeport IN (:worker_1_port, :worker_2_port);

-- pushdown procedures with OUT parameters
SET client_min_messages TO DEBUG1;
CALL proc_pushdown(1, NULL, NULL);
CALL proc_pushdown(1, ARRAY[2000,1], 'AAAA');
```
Once the function is delegated, the INSERT on a distributed table warrants a coordinated transaction, two possibilities here
1) This delegated function is part of the BEGIN/END block at the originator, in such scenarios, the worker will not start a new coordinated transaction, instead uses the one assigned by the coordinator with **assign_distributed_transaction_id()**
2) Delegated function has no surrounding transaction from the coordinator, so it starts its own
**UseCoordinatedTransaction()
      AssignDistributedTransactionId()**

The condition we added is used to distinguish between the two, (1) being unsafe and (2) safe.

As the SELECT is crossing the boundaries of the node, it raises the exception. The below data ensures that the coordinated transaction started locally and returns safely.

![image](https://user-images.githubusercontent.com/44680808/146244736-aa6e9328-f593-4fca-9296-2f953f20b80a.png)

Reply:
You are right. (2) definitely needs an escape hatch here. I wonder whether we should do something similar to what we do for forced calls. That does make things a bit more complex.